### PR TITLE
Avoid IllegalArgumentException without message from ForkJoinPool

### PR DESCRIPTION
**Pull request by jiacheliu3**
Creating ForkJoinPool with unreasonable arguments will lead to IAE without a message, which is confusing. The logic is below:
```
package alluxio.concurrent.jsr;
public ForkJoinPool(int parallelism, ForkJoinWorkerThreadFactory factory,
    UncaughtExceptionHandler handler, boolean asyncMode, int corePoolSize, int maximumPoolSize,
    int minimumRunnable, Predicate<? super ForkJoinPool> saturate, long keepAliveTime,
    TimeUnit unit) {
  // check, encode, pack parameters
  if (parallelism <= 0 || parallelism > MAX_CAP || maximumPoolSize < parallelism
      || keepAliveTime <= 0L)
    throw new IllegalArgumentException();
```

Before:
```
2021-02-04 12:19:29,867 ERROR ProcessUtils - Uncaught exception while running Alluxio master @localhost:19998, stopping it and exiting. Exception "java.lang.IllegalArgumentException", Root Cause "java.lang.IllegalArgumentException"
java.lang.IllegalArgumentException
        at alluxio.concurrent.jsr.ForkJoinPool.<init>(ForkJoinPool.java:757)
        at alluxio.master.AlluxioMasterProcess.startServingRPCServer(AlluxioMasterProcess.java:353)
        at alluxio.master.AlluxioMasterProcess.startServing(AlluxioMasterProcess.java:320)
        at alluxio.master.MasterProcess.startServing(MasterProcess.java:128)
        at alluxio.master.AlluxioMasterProcess.start(AlluxioMasterProcess.java:188)
        at alluxio.ProcessUtils.run(ProcessUtils.java:36)
        at alluxio.master.AlluxioMaster.main(AlluxioMaster.java:55)
```

After:
```
2021-02-06 11:28:42,414 ERROR FaultTolerantAlluxioMasterProcess - Fatal error: Exception thrown in main serving thread
java.lang.IllegalArgumentException: Cannot start Alluxio master gRPC thread pool with alluxio.master.rpc.executor.parallelism=0! The parallelism must be greater than 0!
        at com.google.common.base.Preconditions.checkArgument(Preconditions.java:142)
        at alluxio.master.AlluxioMasterProcess.startServingRPCServer(AlluxioMasterProcess.java:309)
        at alluxio.master.AlluxioMasterProcess.startServing(AlluxioMasterProcess.java:292)
        at alluxio.master.FaultTolerantAlluxioMasterProcess.lambda$gainPrimacy$1(FaultTolerantAlluxioMasterProcess.java:130)
        at java.lang.Thread.run(Thread.java:748)
```
or
```
2021-02-06 11:37:40,880 ERROR FaultTolerantAlluxioMasterProcess - Fatal error: Exception thrown in main serving thread
java.lang.IllegalArgumentException: Cannot start Alluxio master gRPC thread pool with alluxio.master.rpc.executor.parallelism=4 greater than alluxio.master.rpc.executor.max.pool.size=3!
        at com.google.common.base.Preconditions.checkArgument(Preconditions.java:142)
        at alluxio.master.AlluxioMasterProcess.startServingRPCServer(AlluxioMasterProcess.java:313)
        at alluxio.master.AlluxioMasterProcess.startServing(AlluxioMasterProcess.java:292)
        at alluxio.master.FaultTolerantAlluxioMasterProcess.lambda$gainPrimacy$1(FaultTolerantAlluxioMasterProcess.java:130)
        at java.lang.Thread.run(Thread.java:748)
```
or
```
2021-02-06 11:37:40,880 ERROR FaultTolerantAlluxioMasterProcess - Fatal error: Exception thrown in main serving thread
java.lang.IllegalArgumentException: Cannot start Alluxio master gRPC thread pool with alluxio.master.rpc.executor.keepalive=0! The keepalive time must be greater than 0!
        at com.google.common.base.Preconditions.checkArgument(Preconditions.java:142)
        at alluxio.master.AlluxioMasterProcess.startServingRPCServer(AlluxioMasterProcess.java:313)
        at alluxio.master.AlluxioMasterProcess.startServing(AlluxioMasterProcess.java:292)
        at alluxio.master.FaultTolerantAlluxioMasterProcess.lambda$gainPrimacy$1(FaultTolerantAlluxioMasterProcess.java:130)
        at java.lang.Thread.run(Thread.java:748)
```

### CODE REVIEW

**Comment by jiacheliu3** [comment]
@gpang Does this kind of change need to be unit-tested?

File: core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java [hunk 1]
```diff
@@ -305,6 +305,16 @@ protected void startServingRPCServer() {
           GrpcServerAddress.create(mRpcConnectAddress.getHostName(), mRpcBindAddress),
           ServerConfiguration.global(), ServerUserState.global());
 
+      int parallelism = ServerConfiguration.getInt(PropertyKey.MASTER_RPC_EXECUTOR_PARALLELISM);
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java:308]: This is a lot of checking of arguments. Should this go in the constructor, or earlier? This is in the serving method, which secondaries will not even run until much later.

File: core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java [hunk 2]
```diff
@@ -305,6 +305,16 @@ protected void startServingRPCServer() {
...
+      Preconditions.checkArgument(parallelism > 0,
+              String.format("Starting Alluxio master gRPC thread pool with %s executors! "
+                              + "Please set %s to a valid number!",
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java:311]: should we be explicit and say a value greater than 0?

File: core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java [hunk 3]
```diff
@@ -305,6 +305,16 @@ protected void startServingRPCServer() {
           GrpcServerAddress.create(mRpcConnectAddress.getHostName(), mRpcBindAddress),
           ServerConfiguration.global(), ServerUserState.global());
 
+      int parallelism = ServerConfiguration.getInt(PropertyKey.MASTER_RPC_EXECUTOR_PARALLELISM);
+      Preconditions.checkArgument(parallelism > 0,
+              String.format("Starting Alluxio master gRPC thread pool with %s executors! "
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java:310]: It is not actually starting, so should it say something "Cannot start master gRPC thread pool with ..."?

File: core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java [hunk 4]
```diff
@@ -305,6 +305,16 @@ protected void startServingRPCServer() {
...
+                              + "Please set %s to a valid number!",
+                      parallelism, PropertyKey.MASTER_RPC_EXECUTOR_PARALLELISM.toString()));
+      Preconditions.checkArgument(parallelism <= ServerConfiguration.getInt(
+              PropertyKey.MASTER_RPC_EXECUTOR_MAX_POOL_SIZE),
+              String.format("%s cannot be greater than %s!",
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java:315]: Can we show both the name of the parameter and the configured values?

**Review by gpang — CHANGES_REQUESTED** [review]
@jiacheliu3 Thanks for this improvement!

For testing, it seems simple enough for a small unittest which configures the parameters incorrectly, and creating a master process fails?

File: core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java (hunk 1 above)
Comment (jiacheliu3) [reply core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java]: Yea I moved this check up to the `Factory.create()`.

File: core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java [hunk 5]
```diff
@@ -374,6 +373,26 @@ public String toString() {
     return "Alluxio master @" + mRpcConnectAddress;
   }
 
+  private static void validateRPCExecutor() {
+    int parallelism = ServerConfiguration.getInt(PropertyKey.MASTER_RPC_EXECUTOR_PARALLELISM);
+    Preconditions.checkArgument(parallelism > 0,
+            String.format("Cannot start Alluxio master gRPC thread pool with %s=%s! "
+                            + "The parallelism must be greater than 0!",
+                    PropertyKey.MASTER_RPC_EXECUTOR_PARALLELISM.toString(), parallelism));
```
Comment (gpang) [inline core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java:381]: I don't fully recall, but what does this `toString` return? If it returns a lot of other information, then we could use something like `PropertyKey.Name.MASTER_RPC_EXECUTOR_PARALLELISM`?

(this applies to other `toString()` calls in this file)

**Review by gpang — COMMENTED** [review]
@jiacheliu3 Just a minor comment.

File: core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java (hunk 5 above)
Comment (jiacheliu3) [reply core/server/master/src/main/java/alluxio/master/AlluxioMasterProcess.java:381]: The `toString()` for PropertyKey just returns its name, which is exactly `PropertyKey.Name.MASTER_RPC_EXECUTOR_PARALLELISM` when this PropertyKey is created by the Builder. See the sample exception output in the description of this PR.

**Review by gpang — APPROVED** [review]
LGTM

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please

**Pull request merged**